aw_client/client.py

```diff
@@ -72,72 +61,151 @@ def _get(self, endpoint: str) -> Optional[req.Response]:
         data = event.to_json_dict()
-        try:
-            self._post(endpoint, data)
-            self.logger.debug("Sent event to server: {}".format(event))
-        except req.RequestException as e:
-            self.logger.warning("Failed to send event to server ({})".format(e))
-            self._queue_failed_event(bucket, data)
+        if self.connected:
+            try:
+                self._post(endpoint, data)
+                logger.debug("Sent event to server: {}".format(event))
+            except req.RequestException as e:
+                self._connection_lost()
+                logger.warning("Failed to send event to server ({})".format(e))
+                self._queue_failed_request(endpoint, data)
+        else:
+            self._queue_failed_request(endpoint, data)
+
+    #
+    #   Bucket get/post requests
+    #
 
     def get_buckets(self):
         return self._get('buckets').json()
 
-    def get_events(self, bucket) -> List[Event]:
-        endpoint = "buckets/{}/events".format(bucket)
-        events = self._get(endpoint).json()
-        return [Event(**event) for event in events]
+    def setup_bucket(self, bucket_id, event_type: str) -> bool:
+        self.buckets.append({"bid": bucket_id, "etype": event_type})
+        if self.connected:
+            self._create_buckets()
 
-    def create_bucket(self, bucket_id, event_type: str) -> bool:
+    def _create_buckets(self):
         # Check if bucket exists
         buckets = self.get_buckets()
-        if bucket_id in buckets:
-            return False  # Don't do anything if bucket already exists
-        else:
-            # Create bucket
-            endpoint = "buckets/{}".format(bucket_id)
-            data = {
-                'client': self.client_name,
-                'hostname': self.client_hostname,
-                'type': event_type,
-            }
-            try:
-                self._post(endpoint, data)
-            except req.RequestException as e:
-                self.logger.error("Unable to create bucket: {}".format(e))
+        for bucket in self.buckets:
+            if bucket['bid'] in buckets:
+                return False  # Don't do anything if bucket already exists
+            else:
+                # Create bucket
+                endpoint = "buckets/{}".format(bucket['bid'])
+                data = {
+                    'client': self.client_name,
+                    'hostname': self.client_hostname,
+                    'type': bucket['etype'],
+                }
+                try:
+                    self._post(endpoint, data)
+                except req.RequestException as e:
+                    logger.error("Failed to create bucket: {}".format(e))
+                return True
+
+    #
+    #   Failed request queue handling
+    #
+
+    def _queue_failed_request(self, endpoint: str, data: dict):
+        # Find failed queue file
+        entry = {"endpoint": endpoint, "data": data}
+        with self.queue_file_lock:
```

Comment:
If this lock will try to be aquired while _post_failed_requests is being run, this request will still be added to the queue which probably won't be checked for a long time.

A spaghetti way of solving this would be to lock in _try_connect instead and at here check after the lock has been acquired if the connection is back. This would be horribly ugly though so i really don't want do do that.